mops/src/sap/cpu.tpp

```diff
@@ -15,63 +19,82 @@ void mops::sparse_accumulation_of_products(
     Tensor<int32_t, 1> indices_B,
     Tensor<int32_t, 1> indices_output
 ) {
-    if (A.shape[0] != B.shape[0]) {
-        throw std::runtime_error(
-            "A and B tensors must have the same number of elements along the "
-            "first dimension, got " + std::to_string(A.shape[0]) + " and " +
-            std::to_string(B.shape[0])
-        );
-    }
-
-    if (A.shape[0] != output.shape[0]) {
-        throw std::runtime_error(
-            "O must contain the same number of elements as the first "
-            "dimension of A and B , got " + std::to_string(output.shape[0]) +
-            " and " + std::to_string(A.shape[0])
-        );
-    }
-
-    if (C.shape[0] != indices_B.shape[0]) {
-        throw std::runtime_error(
-            "the dimension of C must match that of indices_B, got "
-            + std::to_string(C.shape[0]) +
-            " for C and " + std::to_string(indices_B.shape[0]) + " for indices_B"
-        );
-    }
+    check_sizes(A, "A", 0, B, "B", 0, "sap");
+    check_sizes(A, "A", 0, output, "output", 0, "sap");
+    check_sizes(C, "C", 0, indices_A, "indices_A", 0, "sap");
+    check_sizes(C, "C", 0, indices_B, "indices_B", 0, "sap");
+    check_sizes(C, "C", 0, indices_output, "indices_output", 0, "sap");
+    check_index_tensor(indices_A, "indices_A", A.shape[1], "sap");
+    check_index_tensor(indices_B, "indices_B", B.shape[1], "sap");
+    check_index_tensor(indices_output, "indices_output", output.shape[1], "sap");
 
-    if (C.shape[0] != indices_output.shape[0]) {
-        throw std::runtime_error(
-            "the dimension of C must match that of indices_output, got "
-            + std::to_string(C.shape[0]) +
-            " for C and " + std::to_string(indices_output.shape[0]) + " for indices_output"
-        );
-    }
-
-    // TODO: check sorting (not necessary here, necessary in CUDA implementation)?
-
-    scalar_t* a_ptr = A.data;
-    scalar_t* b_ptr = B.data;
     scalar_t* o_ptr = output.data;
     scalar_t* c_ptr = C.data;
-    int32_t* p_a_ptr = indices_A.data;
-    int32_t* p_b_ptr = indices_B.data;
-    int32_t* p_o_ptr = indices_output.data;
+    int32_t* indices_A_ptr = indices_A.data;
+    int32_t* indices_B_ptr = indices_B.data;
+    int32_t* indices_output_ptr = indices_output.data;
 
     size_t size_first_dimension = A.shape[0];
     size_t size_second_dimension_a = A.shape[1];
     size_t size_second_dimension_b = B.shape[1];
+
     size_t size_second_dimension_o = output.shape[1];
     size_t c_size = C.shape[0];
 
-    for (size_t i = 0; i < size_first_dimension; i++) {
-        size_t shift_first_dimension_a = i * size_second_dimension_a;
-        size_t shift_first_dimension_b = i * size_second_dimension_b;
-        size_t shift_first_dimension_o = i * size_second_dimension_o;
+    std::fill(o_ptr, o_ptr+output.shape[0]*output.shape[1], static_cast<scalar_t>(0.0));
```

Comment:
I think this bunch of code warrants a long comment explaining what is going on with SIMD/interleaving/etc

Currently I have no idea what the code is doing